src/vos/vos_obj_index.c

```diff
@@ -624,6 +624,64 @@ oi_iter_delete(struct vos_iterator *iter, void *args)
 	return rc;
 }
 
+static int
+oi_iter_aggregate(struct vos_iterator *iter, bool discard)
+{
+	struct vos_oi_iter	*oiter = iter2oiter(iter);
+	struct vos_obj_df	*obj;
+	daos_unit_oid_t		 oid;
+	d_iov_t			 rec_iov;
+	bool			 deleted = false;
+	int			 rc;
+
+	D_ASSERT(iter->it_type == VOS_ITER_OBJ);
+
+	d_iov_set(&rec_iov, NULL, 0);
+	rc = dbtree_iter_fetch(oiter->oit_hdl, NULL, &rec_iov, NULL);
+	D_ASSERTF(rc != -DER_NONEXIST,
+		  "Probe should be done before aggregation\n");
+	if (rc != 0)
+		return rc;
+	D_ASSERT(rec_iov.iov_len == sizeof(struct vos_obj_df));
+	obj = (struct vos_obj_df *)rec_iov.iov_buf;
+	oid = obj->vo_id;
+
+	rc = vos_tx_begin(vos_cont2umm(oiter->oit_cont));
+	if (rc != 0)
+		goto exit;
+
+	rc = vos_ilog_aggregate(vos_cont2hdl(oiter->oit_cont), &obj->vo_ilog,
+				&oiter->oit_epr, discard, 0,
+				&oiter->oit_ilog_info);
+	if (rc == 1) {
+		/* Incarnation log is empty, delete the object */
+		D_DEBUG(DB_IO, "Removing object "DF_UOID" from tree\n",
+			DP_UOID(oid));
+		deleted = true;
+		if (!dbtree_is_empty_inplace(&obj->vo_tree)) {
+			/* Keys in subtree are inaccessible */
+			D_DEBUG(DB_IO, "Deleting orphaned subtree\n");
+		}
+		rc = dbtree_iter_delete(oiter->oit_hdl, NULL);
+		D_ASSERT(rc != -DER_NONEXIST);
+	}
+
+	rc = vos_tx_end(vos_cont2umm(oiter->oit_cont), rc);
+exit:
+	if (rc == 0 && deleted) {
+		/* Evict the object from cache */
+		rc = vos_obj_evict_by_oid(vos_obj_cache_current(),
```

Comment:
Can we call vos_obj_evict_by_oid() via vos_oi_delete()? Then nobody will forget to do that in spite of it is triggered by oi_iter_aggregate() or others.